habitat/core/dataset.py

```diff
@@ -236,20 +246,116 @@ def get_splits(
+        the same scene can be loaded consecutively.
+        Sample episodes: sample the specified number of episodes.
+    """
+
+    def __init__(
+        self,
+        episodes: List[T],
+        cycle: bool = True,
+        shuffle_scene_when_cycle: bool = False,
+        group_by_scene: bool = True,
+        max_scene_repeat: int = -1,
+        num_episode_sample: int = -1,
+    ):
+        r"""
+        Initialize episode iterator,
         Args:
-            num_episodes: number of episodes to sample, input -1 to use
-            whole episodes
+            episodes: list of episodes.
+            cycle: if true, cycle back to first episodes when StopIteration.
+            shuffle_scene_when_cycle: if true, shuffle scene groups when cycle.
+            group_by_scene: if true, group episodes from same scene.
+            max_scene_repeat: threshold of how many episodes from the same
+                scene can be loaded consecutively. -1 for no limit
+            num_episode_sample: number of episodes to be sampled.
+                -1 for no sampling.
         """
-        if num_episodes == -1:
-            return
-        if num_episodes < -1:
-            raise ValueError(
-                f"Invalid number for episodes to sample: {num_episodes}"
+        # sample episodes
+        if num_episode_sample != -1:
+            if num_episode_sample < -1:
+                raise ValueError(
+                    f"Invalid episode number to sample: {num_episode_sample}"
+                )
+            episodes = np.random.choice(
+                episodes, num_episode_sample, replace=False
             )
-        self.episodes = np.random.choice(
-            self.episodes, num_episodes, replace=False
-        )
+        self.episodes = episodes
+        self.cycle = cycle
+        self.group_by_scene = group_by_scene
+        if group_by_scene:
+            num_scene_groups = len(
+                list(groupby(episodes, key=lambda x: x.scene_id))
+            )
+            num_unique_scenes = len(set([e.scene_id for e in episodes]))
+            if num_scene_groups >= num_unique_scenes:
+                self.episodes = sorted(self.episodes, key=lambda x: x.scene_id)
+        self.max_scene_repetition = max_scene_repeat
+        self.shuffle_scene_when_cycle = shuffle_scene_when_cycle
+        self.rep_count = 0
+        self._iterator = None
+        self.prev_scene_id = None
+        self._iterator = iter(self.episodes)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        r"""
+        The main logic for handling how episodes will be iterated.
+        Returns:
+            next episode.
+        """
+        try:
+            next_episode = next(self._iterator)
+        except StopIteration:
+            if not self.cycle:
+                raise StopIteration
+            self._iterator = iter(self.episodes)
+            if self.shuffle_scene_when_cycle:
```

Comment:
Should this go before `self._iterator = iter(self.episodes)`?

Reply:
`self._shuffle_iterator_scene()` ideals with iterator directly. The main reason behind that is that I can then use `groupby` function in `itertools`, and I can use this `self._shuffle_iterator_scene()` method for scene switching as well.

Reply:
Maybe make it an `if` and `else`?  Then it'll be clear that `self._iterator = iter(self.episodes)` and `self._shuffle_iterator_scene()` are mutually exclusive

Reply:
hmm.. `self._iterator = iter(self.episodes)` and `self._shuffle_iterator_scene()` aren't actually mutually exclusive atm tho. (It can be if I use `pydash.groupby` that you suggested)

Reply:
Or use `groupby(self.episodes, key=lambda x: x.scene_id)` in `self._shuffle_iterator_scene()`?

Reply:
I also called `self._shuffle_iterator_scene()` [here](https://github.com/JasonJiazhiZhang/habitat-api/blob/9a56168de80418cdd4af4fdf322b2b77ef3983e1/habitat/core/dataset.py#L344) for scene switching when the episode-count for same-scene episode is reached. In this case it's in the middle of an iteration and `self._shuffle_iterator_scene()` shuffles the whatever are still on the iterator, which would be part of `episodes`, possibly not in the original order. I made `self._shuffle_iterator_scene()` dealing with iterator instead of list mainly for that case 😃